src/v/cluster/scheduling/types.h

```diff
@@ -29,78 +31,78 @@ class allocation_state;
  * https://github.com/Netflix/Fenzo/blob/master/fenzo-core/src/main/java/com/netflix/fenzo/VMTaskFitnessCalculator.java
  *
  */
+using replicas_t = std::vector<model::broker_shard>;
+using hard_constraint_evaluator
+  = ss::noncopyable_function<bool(const allocation_node&)>;
+
+using soft_constraint_evaluator
+  = ss::noncopyable_function<uint64_t(const allocation_node&)>;
 
-class hard_constraint_evaluator final {
+class hard_constraint {
 public:
     struct impl {
-        virtual bool evaluate(const allocation_node&) const = 0;
-        virtual void print(std::ostream&) const = 0;
+        virtual hard_constraint_evaluator
+        make_evaluator(const replicas_t&) const = 0;
```

Comment:
nit: kind of hard to understand the meaning of the parameter here, so it would be better if it has a name (`current_replicas` perhaps?)

Reply:
will change